### PR TITLE
[Profiler] Use parameter as key for optimizer state recording.

### PR DESCRIPTION
**Pull request by robieta**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #87568
* #87567
* #87566
* #87006
* #86880
* #86854
* #86853
* #86802
* #87133
* #87244
* #86909
* #86755
* #86754
* __->__ #86753

While optimizer can store state however it likes, in practice most optimizer state corresponds to a particular parameter. (This is the case for all `torch.optim` optimizers.) Thus, it turns out to be ergonomic to collect using that structure. Note that this doesn't lock us into anything; we can always collect state with non Tensor keys if the use case arises.

One simplification that arises is that Module and Optimizer collection has very similar structure. So similar, in fact, that it is possible to use a common template for config. I also found that a lot of the `check_and_store` logic could be simplified and inlined by this joining of collected optimizer state.

Differential Revision: [D40210703](https://our.internmc.facebook.com/intern/diff/D40210703/)

### CODE REVIEW

File: test/profiler/test_profiler.py [hunk 1]
```diff
@@ -1599,29 +1599,34 @@ def _flat_out_extrafields(self, nodes, out=None):
...
                     [(v.storage().data_ptr()) for v in group.get("params", [])],
-                    [(o.storage_data_ptr) for o in opt_.param_addrs]
+                    [(o.storage_data_ptr) for (o, _, _) in opt_.parameters]
                 )
             for opt_ in opts:
-                self.assertEqual(
-                    [(name, val.storage().data_ptr()) for dic in opt.state.values() for name, val in dic.items()],
-                    [(n, p.storage_data_ptr) for (n, p) in opt_.opt_state]
-                )
+                observed_state = {
```
Comment (slgong-fb) [inline test/profiler/test_profiler.py:1621]: Thanks for changing this unit test to work!

File: torch/csrc/profiler/collection.h [hunk 2]
```diff
@@ -238,29 +238,34 @@ using PyMethod = strong_t</*PyMethodDef*/ void*, struct PyMethod_>;
...
   // Indicates that `self_` is the kth instance of `cls_` observed.
   size_t id_{std::numeric_limits<size_t>::max()};
 };
 
 struct OptimizerInfo {
+  struct ParameterInfo {
+    TensorMetadata metadata_;
+    c10::optional<TensorMetadata> grad_metadata_;
+    std::vector<std::pair<std::string, TensorMetadata>> state_;
+  };
```
Comment (slgong-fb) [inline torch/csrc/profiler/collection.h:262]: This is great point. We never meet a state is not matched with a param.

**Review by slgong-fb — APPROVED** [review]

File: test/profiler/test_profiler.py [hunk 3]
```diff
@@ -1599,29 +1599,34 @@ def _flat_out_extrafields(self, nodes, out=None):
...
+                    p.storage_data_ptr: {name: s.storage_data_ptr for name, s in state}
+                    for (p, _, state) in opt_.parameters
+                }
+                for parameter, parameter_state in opt.state.items():
+                    self.assertEqual(
+                        {name: value.storage().data_ptr() for name, value in parameter_state.items()},
+                        observed_state.get(parameter.storage().data_ptr(), [])
+                    )
```
Comment (aaronenyeshi) [inline test/profiler/test_profiler.py:1629]: Would it be possible to add a comment for this assert? I am getting confused what it is checking. It looks like its verifying that opt_.parameters.state.storage_data_ptr is equal to parameter_state.value.storage.data_ptr?

File: torch/csrc/autograd/profiler_python.cpp [hunk 4]
```diff
@@ -204,20 +205,36 @@ struct Config<CallType::PyCall> {
...
+  using cls_t = Cls;
   using ephemeral_t = PyFrameObject*;
-  using info_t = std::pair<cls_t, std::vector<ParameterInfo>>;
-  struct cache_t {
-    c10::optional<CodeLocation> location_; // nn.Module.forward;
-    ska::flat_hash_map<key_t, info_t> modules_and_params_;
+
+  struct ClsAndParameters {
+    cls_t cls_;
+    std::vector<typename Info::ParameterInfo> parameters_;
```
Comment (aaronenyeshi) [inline torch/csrc/autograd/profiler_python.cpp:216]: For optimizer, are the states stored inside parameters_ now?

File: torch/csrc/autograd/profiler_python.cpp [hunk 5]
```diff
@@ -204,20 +205,36 @@ struct Config<CallType::PyCall> {
   static constexpr EventType event_type = EventType::PyCall;
 };
 
-template <>
-struct Config<CallType::PyModuleCall> {
-  using key_t = PyModuleSelf;
-  using cls_t = PyModuleCls;
+template <typename Key, typename Cls, typename Info>
```
Comment (aaronenyeshi) [inline torch/csrc/autograd/profiler_python.cpp:208]: Why not directly use typename ParameterInfo(which has the key, cls, and parameter info)? Or replace typename Info with typename ParameterInfo?

**Review by aaronenyeshi — APPROVED** [review]
This looks great! The code looks more extensible and clean!

File: test/profiler/test_profiler.py (hunk 3 above)
Comment (robieta) [reply test/profiler/test_profiler.py:1629]: Will do.

File: torch/csrc/autograd/profiler_python.cpp (hunk 4 above)
Comment (robieta) [reply torch/csrc/autograd/profiler_python.cpp:216]: Correct.

File: torch/csrc/autograd/profiler_python.cpp (hunk 5 above)
Comment (robieta) [reply torch/csrc/autograd/profiler_python.cpp:208]: That's a great point.

**Pull request closed without merging**